api/server.go

```diff
@@ -236,6 +237,58 @@ func (s *server) consensusUpdatesIndexHandler(jc jape.Context) {
 	jc.Encode(res)
 }
 
+func (s *server) miningGetBlockTemplateHandler(jc jape.Context) {
+	var req MiningGetBlockTemplateRequest
+	if jc.Decode(&req) != nil {
+		return
+	}
+
+	// TODO: add polling
+
+	template, err := generateBlockTemplate(s.cm, req.PayoutAddress)
+	if jc.Check("failed to generate block template", err) != nil {
+		return
+	}
+	jc.Encode(template)
+}
+
+func (s *server) miningSubmitBlockTemplateHandler(jc jape.Context) {
+	var req MiningSubmitBlockRequest
+	rawBlock, err := hex.DecodeString(req.Params[0])
```

Comment:
Is there a reason we can't do:

```go
var req MiningSubmitBlockRequest
f jc.Decode(&req) != nil {
    return
} else if len(req.Params) < 1 {
    jc.Error(errors.New("expected block hex in request params array"), http.StatusBadRequest)
    return
}

// hex decode block
rawBlock, err := hex.DecodeString(req.Params[0])
if jc.Check("couldn't decode block", dec.Err()) != nil {
    return
}

// decode block
...
```
Also, we could consider return a 400 if `len(req.Params)` is `!= 1`, which is better imo.

Reply:
A lot of this is essentially copied and kept for compatibility. e.g. BIP22 allows for any number of params to be sent and if a pool uses some piece of code that attaches more information I don't want to return an error here.